c++/src/ColumnPrinter.cc

```diff
@@ -252,7 +256,7 @@ namespace orc {
       break;
 
     case STRUCT:
-      result = new StructColumnPrinter(buffer, type);
+      result = new StructColumnPrinter(buffer, type, selectedColumns);
```

Comment:
You also need to push the selected columns down through the list, map, and union types. Otherwise, you won't be able to select columns below them.